File: Chapter 3/Section 3.1/humble/humble.py
"""
ID: ten.to.1
TASK: humble
LANG: PYTHON3
"""

#!/usr/bin/python

#Help from zzh8829
#https://usacosolutions.wordpress.com/2012/09/30/usaco-3-1-prog-humble/

from itertools import combinations_with_replacement
import bisect

file_in  = open("humble.in", "r")
file_out = open("humble.out", "w")

K, N   = map(int, file_in.readline().split())
P      = list(map(int, file_in.read().split()))
S      = P.copy()
values = { }

for p in P:
	values[p] = True

for p in P:
	for n in range(0, N):
		new_num = p * S[n]
		
		if(values.get(new_num)):
			continue
		
		if(len(S) < N):
			bisect.insort(S, new_num)
			values[new_num] = True
		elif(new_num < S[-1]):
			bisect.insort(S, new_num)
			values[new_num] = True


file_out.write(str(S[N - 1]) + "\n")

File: Chapter 3/Section 3.1/humble/humble.cpp
/*
ID: ten.to.1
TASK: humble
LANG: C++
*/

//Implemenation from zzh8829
//https://usacosolutions.wordpress.com/2012/09/30/usaco-3-1-prog-humble/

#include <iostream>
#include <fstream>
#include <set>
 
int P[101];
std::set<int> S;
 
int main()
{
	std::ifstream file_in("humble.in");
	std::ofstream file_out("humble.out");
	
	int K, N; file_in >> K >> N;
	for(int i = 0; i != K ;i++)
	{
		file_in >> P[i];
		S.insert(P[i]);
	}
	
	//Go through each prime once
	for(int i = 0; i != K; i++)
	{
		std::set<int>::iterator S_it = S.begin();
		
		//Keep looping through humble numbers
		while(true)
		{
			//t == new_num
			int t = (*S_it) * P[i];
			
			//Avoid overflow
			if(t < 0)
				break;
			
			if(S.size() > N)
			{
				//Delete last one
				//-- Increment Before then execute command
				S.erase(--S.end());
				
				if(t > (*(--S.end())))
					break;
			}
			
			//Insert only if t is less than max range
			S.insert(t);
			S_it++;
		}
	}
	
	file_out << *(--S.end()) << '\n';
}
